### ISSUE TITLE
Investigate implications of "burst checking" behavior

### ISSUE THREAD

**Issue by vito** [concourse/concourse#4624]
LIDAR kicks off all the queued `check`s at once, which is different from Radar where each resource had its own interval. Theoretically this would mean spikier metrics.

What is the impact of this? Can/should we do something to mitigate it?

**Comment by vito** [comment]
Note: the super spiky metrics started after https://github.com/concourse/concourse/pull/4573 - maybe we should make it notify...fewer things?

**Comment by pivotal-jwinters** [comment]
basically same as https://github.com/concourse/concourse/issues/4662.

overall metrics seem to have improved but we'll continue to monitor things.

**Comment by vito** [comment]
Reopening per https://github.com/concourse/concourse/issues/5360#issuecomment-613660843

Some ideas in https://github.com/concourse/concourse/issues/5360#issuecomment-613663399 and https://github.com/concourse/concourse/issues/5360#issuecomment-613676006

**Comment by pivotal-jwinters** [comment]
@vito this is probably also further proof that #3424 might not be a good idea.

**Comment by vito** [comment]
@pivotal-jwinters Do you have more thoughts on that? Not seeing the immediate connection. 🤔

**Comment by pivotal-jwinters** [comment]
@vito https://github.com/concourse/concourse/issues/5360#issuecomment-606820787 shows spiky CPU on the workers. And that's when we're reusing the check containers. I can only imagine that the problem gets worse if we're creating those containers from scratch every time.

**Comment by vito** [comment]
Probably, but I think we'd have to measure the performance of concurrent container creation to really be sure, and then see what options we have to improve it if necessary. Container creation isn't *intrinsically* expensive - it's just spawning a process with flags - so if we can minimize the costly aspects (i.e. preparing the filesystem and network) that would probably help a lot. We already have to deal with creating lots of containers when running lots of builds, and it doesn't seem to be a big problem there. (We run `check`s a lot more frequently but we also have folks who run a ton of builds that create a ton of containers.)

We may also end up reaching for broader mitigations such as backing off for infrequently-changing resources (https://github.com/concourse/concourse/issues/3975) so we don't have to do as much checking in the first place. The other half of the problem is the load on external services and the local resource usage of the actual `check`ing, which seems more important anyway. Between the two problems reported (https://github.com/concourse/concourse/issues/5360#issuecomment-606820787 vs. https://github.com/concourse/concourse/issues/5360#issuecomment-609424802), the only real problem with the CPU usage was that metrics went off, while the problem with the load on external services required human intervention to fix.

**Comment by cirocosta** [comment]
> while the problem with the load on external services required human intervention to fix.

reminded me that this might be even more of a thing on those systems that rate limit based on [token-bucket](https://en.wikipedia.org/wiki/Token_bucket) schemes w/out allowance for high burstiness

**Comment by jchesterpivotal** [comment]
Yes, a token scheme is probably sound. Another alternative is to load-level the checks over (say) one minute windows. But I prefer token-bucket: it would be better and easier to provide operator knobs for.

I'll also bring up some classics from my endless laundry list of ideas I don't do any work on: jittered checks (resource scheduling happens with slight randomisation of delay to spread out load) and splitting out a versions API (versions are inserted into Concourse by workloads it doesn't need to manage).

**Comment by clarafu** [comment]
After talking with @vito, we came up with a possible plan for fixing this at least in the short-term.

1. Do some isolated simulation with metrics set up to observe the high spike in checks that happen with the way that lidar is currently configured. We should see that many checks are fired off all at the same time.
2. Implement a max in flight similar to the build scheduler.
3. Run the same isolated simulation and observe that the curve is flattened and there is no longer a huge spike of checks all at once, but rather more spread out.

**Comment by christophermancini** [comment]
Question, if we have a default check interval for 5 minutes, but a webhook triggers a check, does it restart the 5min timer?

**Comment by pivotal-jwinters** [comment]
@christophermancini anything that triggers a check (webhook, manual trigger) will restart the timer.

**Comment by clarafu** [comment]
I created an application that will simulate resource checking by using the lidar component and faking out all of it's dependencies (database and runtime portions). You can find it here https://github.com/clarafu/lidar-sim.

I ran the simulation and hooked up tracing in order to view metrics through kibana. There are 100 resources configured in this simulation and each check is faked out to complete in 1 second. I ran it twice, once using the lidar component from master and once with `max-in-flight` implemented in lidar. The first screenshot shows the number of checks over a period of 5 minutes with `max-in-flight` not implemented.

![Screen Shot 2020-05-27 at 4 38 35 PM](https://user-images.githubusercontent.com/19290703/83071086-89488500-a03a-11ea-839f-3cd289f43d16.png)

All 100 checks happen at once on an interval. This is because the `checker` component within lidar grabs all the started checks from the checks table and then runs the check for each within it's own goroutine. 

When I ran the same simulation with the `max-in-flight` implemented (with it set to only allow 10 checks to run at a time), you can observe how the peak is only 10 checks and the actual execution of all the checks are more spread out.

![Screen Shot 2020-05-27 at 4 45 11 PM 1](https://user-images.githubusercontent.com/19290703/83071709-8d28d700-a03b-11ea-8ce5-c80936f42d9c.png)

A downside of adding a `max-in-flight` implementation to the checker component would be that it takes longer to complete all the checks. If you take a look at the time it takes for a full check process to complete (which includes a resource scan from the scanner and then being checked by the checker), without the `max-in-flight` it took approximately 20 seconds to complete but with a `max-in-flight` implementation set to 10 checks at a time, it takes 30 seconds to complete.

**Comment by clarafu** [comment]
Another thing I looked into was the impact of a lower interval for the scanner. The scanner currently runs every 1 minute by default and if we were to simulate a real environment with 2000 resources and the max in flight set to 32, you can see that the number of checks that happen over a time period is still quite spiky.

![Screen Shot 2020-05-29 at 3 49 13 PM](https://user-images.githubusercontent.com/19290703/83303649-72856800-a1cb-11ea-8698-560c5141cc49.png)

First off, if you just look at the graph it might be misleading because it looks like 700 checks are running at once but that is not supposed to be possible with the max in flight of 32. It actually only looks like it is running 700 checks at once but actually the graph is just grouping together the number of checks because the x axis is per 5 seconds. This means that in 5 seconds of the checker running, 700 checks were run within the max in flight of 32 checks at once.

So looking at this graph, even if we do have a max in flight set it seems like the checks are still spiky. I think this might be a result of the lidar scanner component only running every 1 minute, so the number of resources that need to be checked build up within that 1 minute interval and when the scanner runs, it creates checks for all those resources. So I tried lowering the interval and saw that the checks seem to balance out more. It seems like total number of checks that run within one interval of the checker is lower because the checks are more spread out between multiple intervals of the checker.

![Screen Shot 2020-05-29 at 4 12 07 PM](https://user-images.githubusercontent.com/19290703/83303675-7a450c80-a1cb-11ea-874a-4d73bc485e2d.png)

I think the scanner was set to run every minute because radar used to run on a 1 minute interval. This was probably set as the same number possibly because the resource-checking-interval is set to a default of 1 minute and there is not much point in having the scanner check more often than the resource's need to be checked. But if we have the scanner interval to be lower than the resource checking interval, it can help lower the amount of checks scheduled per interval tick of the scanner->checker.

**Comment by jamieklassen** [comment]
@clarafu can you share how you are producing these kibana charts? I'm guessing it's an area chart where the source data is spans where `operationName : checker.Run`, the Y-axis is count and the X-axis is a date histogram on `startTimeMillis`. Is this correct?

**Comment by clarafu** [comment]
@jamieklassen Haha ya that's exactly it 😆

**Comment by vito** [comment]
Here's how this looks on Hush House, before:

![image](https://user-images.githubusercontent.com/1880/84416707-90e35e80-abe2-11ea-81cc-e050e50ef1f3.png)

And after:

![image](https://user-images.githubusercontent.com/1880/84416746-9d67b700-abe2-11ea-8ed7-4941345d57c0.png)

🎉
